File: CSharpFundamentals/CSharpOOPbasics/PolymorphismEx/VehiclesExtension/Bus.cs
namespace Vehicles
{
    public class Bus : Vehicle
    {
        private const double A_C_CONSUMPTION_CAR_FULL = 1.4;
        private const double A_C_CONSUMPTION_CAR_EMPTY = 0;
        public bool IsEmpty { get; set; }

        public Bus(double fuelQuantity, double fuelConsumption, double tankCapacity)
            : base(fuelQuantity, fuelConsumption, tankCapacity)
        {
        }

        public override void Drive(double distance)
        {
            if (IsEmpty)
            {
                this.FuelConsumption += A_C_CONSUMPTION_CAR_EMPTY;
                base.Drive(distance);
                this.FuelConsumption -= A_C_CONSUMPTION_CAR_EMPTY;
            }
            else
            {
                this.FuelConsumption += A_C_CONSUMPTION_CAR_FULL;
                base.Drive(distance);
                this.FuelConsumption -= A_C_CONSUMPTION_CAR_FULL;
            }
        }
    }
}

File: JavaFundamentals/JavaOOP/Polymorphism-Ex/src/vehicles_extension/Bus.java
package vehicles_extension;

import java.text.DecimalFormat;

public class Bus extends Vehicle {
    private static final Double AC_FUEL_INCREASE_BUS = 1.4;

    public Bus(Double fuelQuantity, Double fuelConsumption, Double tankCapacity) {
        super(fuelQuantity, fuelConsumption, tankCapacity);
    }

    @Override
    public boolean isValidDistance(Double distance) {
        return distance <= this.getFuelQuantity() / (this.getFuelConsumption() + AC_FUEL_INCREASE_BUS);
    }

    @Override
    public void drive(Double distance) {
        if (isValidDistance(distance)) {
            DecimalFormat format = new DecimalFormat("#.##");
            Double consumption = this.getFuelConsumption() + AC_FUEL_INCREASE_BUS;
            Double fuel = this.getFuelQuantity() - distance * consumption;
            this.setFuelQuantity(fuel);
            System.out.println(String.format("Bus travelled %s km", format.format(distance)));
        } else {
            System.out.println("Bus needs refueling");
        }
    }

    public void driveEmpty(Double distance) {
        if (isValidDistance(distance)) {
            DecimalFormat format = new DecimalFormat("#.##");
            Double consumption = this.getFuelConsumption();
            Double fuel = this.getFuelQuantity() - distance * consumption;
            this.setFuelQuantity(fuel);
            System.out.println(String.format("Bus travelled %s km", format.format(distance)));
        } else {
            System.out.println("Bus needs refueling");
        }
    }

    @Override
    public void refuel(Double fuelAmount) {
        if (fuelAmount <= 0 ) {
            throw new IllegalArgumentException("Fuel must be a positive number");
        }
        this.setFuelQuantity(this.getFuelQuantity() + fuelAmount);
    }
}
